File: 40CombinationII.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:

        res = []

        def backtrack(arr, start, target):
            if target == 0:
                res.append(arr[:])
                return
            for i in range(start, len(candidates)):
                if candidates[i] > target:
                    continue
                if i > start and candidates[i] == candidates[i - 1]:
                    continue
                arr.append(candidates[i])
                backtrack(arr, i + 1, target - candidates[i])
                arr.pop()

        candidates.sort()
        backtrack([], 0, target)
        return res


sol = Solution()
print(sol.combinationSum2([10, 1, 2, 7, 6, 1, 5], 8))

File: 40CombinationII.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        
        backtrack(res, new ArrayList<>(), candidates, target, 0);
        return res;
    }
    private static void backtrack(List<List<Integer>> res, List<Integer> arr, int[] candidates, int remain, int start) {
        if (remain == 0) {
            res.add(new ArrayList<>(arr));
            return;
        }
        for (int i = start; i < candidates.length; i++) {
            if (candidates[i] > remain) {
                break;
            }
            arr.add(candidates[i]);
            backtrack(res, arr, candidates, remain - candidates[i], i);
            arr.remove(arr.size() - 1); 
        }
    }
    public static void main(String[] args) {
        int[] nums = {1, 2, 3, 1};

        System.err.println(combinationSum(nums, 8));
    }
}